File: Core/Workspace/CSharp/Adapters/Allors.Workspace.Adapters.Direct/Database/DatabaseObject.cs
// <copyright file="DatabaseObject.cs" company="Allors bvba">
// Copyright (c) Allors bvba. All rights reserved.
// Licensed under the LGPL license. See LICENSE file in the project root for full license information.
// </copyright>

namespace Allors.Workspace.Adapters.Direct
{
    using System;
    using System.Collections.Generic;
    using Meta;
    using System.Linq;
    using Security;

    public class DatabaseObject
    {
        private Permission[] deniedPermissions;
        private AccessControl[] accessControls;

        private Dictionary<Guid, object> roleByRelationTypeId;
        //private SyncResponseRole[] syncResponseRoles;

        internal DatabaseObject(WorkspaceDatabase workspaceDatabase, long databaseId, IClass @class)
        {
            this.WorkspaceDatabase = workspaceDatabase;
            this.DatabaseId = databaseId;
            this.Class = @class;
            this.Version = 0;
        }

        internal DatabaseObject(WorkspaceDatabase workspaceDatabase)
        {
            this.WorkspaceDatabase = workspaceDatabase;
            //this.DatabaseId = long.Parse(syncResponseObject.I);
            //this.Class = (IClass)this.WorkspaceDatabase.MetaPopulation.Find(Guid.Parse(syncResponseObject.T));
            //this.Version = !string.IsNullOrEmpty(syncResponseObject.V) ? long.Parse(syncResponseObject.V) : 0;
            //this.syncResponseRoles = syncResponseObject.R;
            //this.SortedAccessControlIds = ctx.ReadSortedAccessControlIds(syncResponseObject.A);
            //this.SortedDeniedPermissionIds = ctx.ReadSortedDeniedPermissionIds(syncResponseObject.D);
        }

        public WorkspaceDatabase WorkspaceDatabase { get; }

        public IClass Class { get; }

        public long DatabaseId { get; }

        public long Version { get; private set; }

        public string SortedAccessControlIds { get; }

        public string SortedDeniedPermissionIds { get; }

        private Dictionary<Guid, object> RoleByRelationTypeId
        {
            get
            {
                //if (this.syncResponseRoles != null)
                //{
                //    var metaPopulation = this.WorkspaceDatabase.MetaPopulation;
                //    this.roleByRelationTypeId = this.syncResponseRoles.ToDictionary(
                //        v => Guid.Parse(v.T),
                //        v =>
                //        {
                //            var value = v.V;
                //            var RoleType = ((IRelationType)metaPopulation.Find(Guid.Parse(v.T))).RoleType;

                //            var objectType = RoleType.ObjectType;
                //            if (objectType.IsUnit)
                //            {
                //                return UnitConvert.Parse(RoleType.ObjectType.Id, value);
                //            }
                //            else
                //            {
                //                if (RoleType.IsOne)
                //                {
                //                    return value != null ? long.Parse(value) : (long?)null;
                //                }
                //                else
                //                {
                //                    return value != null
                //                        ? value.Split(Encoding.SeparatorChar).Select(long.Parse).ToArray()
                //                        : Array.Empty<long>();
                //                }
                //            }
                //        });

                //    this.syncResponseRoles = null;
                //}

                return this.roleByRelationTypeId;
            }
        }

        private AccessControl[] AccessControls => throw new NotImplementedException();

        private Permission[] DeniedPermissions => throw new NotImplementedException();

        public object GetRole(IRoleType roleType)
        {
            object @object = null;
            this.RoleByRelationTypeId?.TryGetValue(roleType.RelationType.Id, out @object);
            return @object;
        }

        public bool IsPermitted(Permission permission) =>
            permission != null &&
            !this.DeniedPermissions.Contains(permission) &&
            this.AccessControls.Any(v => v.PermissionIds.Any(w => w == permission.Id));
    }
}

File: Core/Workspace/Typescript/libs/workspace/adapters/memory/src/lib/Database/DatabaseObject.ts
import { RoleType } from '@allors/workspace/meta/system';

export class DatabaseObject {
  deniedPermissions: Permission[];
  accessControls: AccessControl[];

  roleByRelationType: Map<RelationType, unknown>;
  syncResponseRoles: SyncResponseRole[];

  constructor(public readonly database: Database, public readonly identity: number, public readonly cls: Class) {
    this.version = 0;
  }

  // internal DatabaseObject(Database database, ResponseContext ctx, SyncResponseObject syncResponseObject)
  // {
  //     this.Database = database;
  //     this.Identity = syncResponseObject.Id;
  //     this.Class = (IClass)this.Database.MetaPopulation.FindByTag(syncResponseObject.ObjectType);
  //     this.Version = syncResponseObject.Version;
  //     this.syncResponseRoles = syncResponseObject.Roles;
  //     this.AccessControlIds = syncResponseObject.AccessControls != null ? (ISet<long>)new HashSet<long>(ctx.CheckForMissingAccessControls(syncResponseObject.AccessControls)) : EmptySet<long>.Instance;
  //     this.DeniedPermissionIds = syncResponseObject.DeniedPermissions != null ? (ISet<long>)new HashSet<long>(ctx.CheckForMissingPermissions(syncResponseObject.DeniedPermissions)): EmptySet<long>.Instance;
  // }

  version: number;

  accessControlIds: Set<number>;

  deniedPermissionIds: Set<number>;

  get roleByRelationType(): Map<RelationType, object> {
    // if (this.syncResponseRoles != null)
    // {
    //     var meta = this.database.metaPopulation;
    //     var metaPopulation = this.database.MetaPopulation;
    //     this.roleByRelationType = this.syncResponseRoles.ToDictionary(
    //         v => (IRelationType)meta.FindByTag(v.RoleType),
    //         v =>
    //         {
    //             var roleType = ((IRelationType)metaPopulation.FindByTag(v.RoleType)).RoleType;
    //             var objectType = roleType.ObjectType;
    //             if (objectType.IsUnit)
    //             {
    //                 return UnitConvert.FromString(roleType.ObjectType.Tag, v.Value);
    //             }
    //             if (roleType.IsOne)
    //             {
    //                 return v.Object;
    //             }
    //             return v.Collection;
    //         });
    //     this.syncResponseRoles = null;
    // }
    // return this.roleByRelationType;
  }

  get AccessControls(): AccessControl[] {
    // return this.accessControls switch
    // {
    //     null when this.AccessControlIds == null => Array.Empty<AccessControl>(),
    //     null => this.AccessControlIds
    //         .Select(v => this.Database.AccessControlById[v])
    //         .ToArray(),
    //     _ => this.accessControls
    // };
  }

  get deniedPermissions(): Permission[] {
    // this.deniedPermissions = this.deniedPermissions switch
    // {
    //     null when this.DeniedPermissionIds == null => Array.Empty<Permission>(),
    //     null => this.DeniedPermissionIds
    //         .Select(v => this.Database.PermissionById[v])
    //         .ToArray(),
    //     _ => this.deniedPermissions
    // };
  }

  getRole(roleType: RoleType): unknown {
    // object @object = null;
    // _ = this.RoleByRelationType?.TryGetValue(roleType.RelationType, out @object);
    // return @object;
  }

  isPermitted(permission: Permission): boolean {
    permission != null && !this.DeniedPermissions.Contains(permission) && this.AccessControls.Any((v) => v.PermissionIds.Any((w) => w == permission.Id));
  }

  updateDeniedPermissions(updatedDeniedPermissions: number[]) {
    if (this.deniedPermissions == null) {
      this.deniedPermissionIds = [];
    } else {
      // if (!this.deniedPermissionIds.SetEquals(updatedDeniedPermissions))
      // {
      //     this.DeniedPermissionIds = new HashSet<long>(updatedDeniedPermissions);
      // }
    }
  }
}
